modules/data.remote/R/call_MODIS.R

```diff
@@ -27,65 +29,116 @@
   size <- 0
+# reformat start and end date if they are in YYYY/MM/DD format instead of YYYYJJJ
+  if (grepl("/", start_date) == T)
+  {
+    start_date = as.Date(paste0(lubridate::year(start_date), spatial.tools::add_leading_zeroes(lubridate::yday(start_date), 3)), format = "%Y%j")
+  }
   
-  # set start and end dates to correct format
-  if (package_method == "MODISTools"){
-    
+  if (grepl("/", end_date) == T)
+  {
+    end_date = as.Date(paste0(lubridate::year(end_date), spatial.tools::add_leading_zeroes(lubridate::yday(end_date), 3)), format = "%Y%j")
+  }
+  start_date = as.Date(start_date, format = "%Y%j")
+  end_date = as.Date(end_date, format = "%Y%j")
+
+
+  #################### if package_method == MODISTools option ####################
+  if (package_method == "MODISTools")
+  {
+    #################### FUNCTION PARAMETER PRECHECKS #################### 
+    #1. check that modis product is available
     products = MODISTools::mt_products()
     if (!(product %in% products$product))
     {
       print(products)
       stop("Product not available for MODIS API. Please chose a product from the list above.")
-    } else {
-      print("Check #1: Product exists!")
-    }
-    
-
-    dates <- MODISTools::mt_dates(product = product, lat = lat, lon = lon)$modis_date
-    dates <- as.numeric(substr(dates, 2, nchar(dates)))
-    if (as.numeric(start_date) <= dates[1] | as.numeric(end_date) >= dates[length(dates)])
-    {
-      print(paste("Range of dates for product are ", dates[1], " - ", dates[length(dates)], sep = ""))
-      stop("Please choose dates between the date range listed above.")
-    } else {
-      print("Check #2: Dates are available!")
-    }
+    } 
     
+    #2. check that modis produdct band is available
     bands <- MODISTools::mt_bands(product = product)
     if (!(band %in% bands$band))
     {
       print(bands$band)
       stop("Band selected is not avialable. Please selected from the bands listed above that correspond with the data product.")
-    } else {
-      print("Check #3: Band Exists!")
+    } 
+    
+    #3. check that dates asked for in function parameters are fall within dates available for modis product/bands.
+    dates <- MODISTools::mt_dates(product = product, lat = lat, lon = lon)$modis_date
+    dates <- as.Date(as.character(substr(dates, 2, nchar(dates))), format = "%Y%j")
+    
+    ########## Date case 1:  user only wants one date ##########
+    if (start_date == end_date)
+    {
+      if (as.numeric(start_date) >= dates[1] & as.numeric(end_date) <= dates[length(dates)])
+      {
+        print("Extracting data")
+        
+        start <- as.Date(start_date, "%Y%j")
+        end <- as.Date(end_date, "%Y%j")
+      }
+      ########## For Date case 1: if only one date is asked for, but date is not within modis data prodate date range ##########
+      if (as.numeric(start_date)< dates[1] | as.numeric(end_date)> dates[length(dates)])
```

Comment:
Inside of `if` statements, you _always_ want to be using the "double" versions of R's "and/or" operators (i.e. `&&` or `||`).

The single character forms -- `&` and `|` -- are vectorized over their inputs, so they return a vector. `if` doesn't know what to do with a logical vector, so if it gets a vector, it will throw a warning and use only the first value.

``` r
x <- sample(1:3, 10, TRUE)
y <- sample(1:2, 10, TRUE)
x == 1 & y == 2
#>  [1] FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
if (x == 1 & y == 2) {
  print("yes")
} else {
  print("no")
}
#> Warning in if (x == 1 & y == 2) {: the condition has length > 1 and only
#> the first element will be used
#> [1] "no"
```

The double character forms (`&&` and `||`) evaluate strictly from left to right, stopping as early as possible and returning a scalar result.

```r

x == 1 && y == 2
#> [1] FALSE
FALSE && stop("This is an error")
#> [1] FALSE
```

If you are doing stuff with `if` and vectors, you generally will want to use the `any` and/or `all` functions. E.g.

```r
if (any(x == 3) && all(y > 0)) { # do stuff...
```

Reply:
Fixed all the doubling up for | and &. Unless there is a good reason I don't know about the any() and all() functions seem like overkill for conditional statements.